lib/spack/spack/binary_distribution.py

```diff
@@ -253,6 +257,9 @@ def find_by_hash(self, find_hash, mirrors_to_check=None):
             mirrors_to_check: Optional mapping containing mirrors to check.  If
                 None, just assumes all configured mirrors.
         """
+        if find_hash not in self._mirrors_for_spec:
+            # Not found in the cached index, pull the latest from the server.
+            self.update(with_cooldown=True)
```

Comment:
This bothers me too about this PR. Please do not make read operations do write operations.

Reply:
Caches as a general rule do a write operation on cache miss. That's what's happening here. This should be normal and I was very surprised to find it not doing something like this before.

Reply:
What you wrote is more wrong than the previous implementation was wrong